sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,210 @@
+     *     If it is a success the value in the cache will be updated. If the refresh task throws an exception the key will be removed from the cache.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function failed - new one will be launched.
+     * </p>
+     *
+     * @param key Key for which to get a value.
+     * @param singleValueInitFunc Initialization function.
+     * @param forceRefresh Force refresh for refreshing the cache
+     * @return Cached value or value returned by initialization function.
+     */
+    public Mono<TValue> getAsync(
+        TKey key,
+        Function<TValue, Mono<TValue>> singleValueInitFunc,
+        Function<TValue, Boolean> forceRefresh) {
+
+        AsyncLazyWithRefresh<TValue> initialLazyValue = values.get(key);
+        if (initialLazyValue != null) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("cache[{}] exists", key);
+            }
+
+            return initialLazyValue.getValueAsync().flatMap(value -> {
+                if(!forceRefresh.apply(value)) {
+                    return Mono.just(value);
+                }
+
+                Mono<TValue> refreshMono = initialLazyValue.createAndWaitForBackgroundRefreshTaskAsync(singleValueInitFunc);
+
+                return refreshMono.onErrorResume(
+                    (exception) -> {
+                        if (logger.isDebugEnabled()) {
+                            logger.debug("refresh cache [{}] resulted in error", key, exception);
+                        }
+                        if (initialLazyValue.shouldRemoveFromCache()) {
+                            this.remove(key);
+                        }
+                        return Mono.error(exception);
+                    }
+                );
+            }).onErrorResume((exception) -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("cache[{}] resulted in error", key, exception);
+                }
+                if (initialLazyValue.shouldRemoveFromCache()) {
+                    this.remove(key);
+                }
+                return Mono.error(exception);
+            });
+        }
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("cache[{}] doesn't exist, computing new value", key);
+        }
+        AsyncLazyWithRefresh<TValue> asyncLazyWithRefresh = new AsyncLazyWithRefresh<TValue>(singleValueInitFunc);
+        this.values.putIfAbsent(key, asyncLazyWithRefresh);
+        AsyncLazyWithRefresh<TValue> result = this.values.get(key);
+
+        return result.getValueAsync().onErrorResume(
+            (exception) -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("cache[{}] resulted in error", key, exception);
+                }
+                // Remove the failed task from the dictionary so future requests can send other calls..
+                this.remove(key);
```

Comment:
we probably should also use shouldRemoveFromCache() here. else there is a potential race condition:
Thread 1 -> get the valueMono
Thread 2 -> get the valueMono
valueMono -> failed, thread 1 reach here to remove the key
Thread 3 -> start a new valueMono
Thread 2 reach to the error handling and remove the key, only this time it removed the wrong one ?

Reply:
Added the check here as well.